### ISSUE TITLE
Speed bottleneck when accumulating progress_bar_metrics

### ISSUE THREAD

**Issue by tobiasfshr** [PyTorchLightning/pytorch-lightning#12976]
## 🐛 Bug

Since trainer.progress_bar_metrics copies to cpu (via metrics_to_scalars calling .item(), compare L591 of trainer/connectors/logger_connector/result.py), it is not possible to accumulate and average progress bar metrics inside a custom ProgressBar without creating a speed bottleneck due to copying to cpu after every training iteration.

### To Reproduce

Log some scalar tensors, call trainer.progress_bar_metrics after each batch in your custom progress bar, observe speed bottleneck (especially in multi-GPU training).

### Expected behavior

Checking that progress bar metrics are actually scalars and copying to cpu are two different things. trainer.progress_bar_metrics could return scalar tensors, which can be copied to cpu in the ProgressBar as needed (i.e. only when printing).

cc @carmocca @edward-io @ananthsub @rohitgr7 @kamil-kaczmarek @Raalsky @Blaizzy @kaushikb11 @awaelchli @borda @akihironitta

**Comment by carmocca** [comment]
Hi!

We could do your suggested change. But it would be a breaking change for any other component that relies on the values of `progress_bar_metrics` being simple floats.

> it is not possible to accumulate and average progress bar metrics inside a custom ProgressBar without creating a speed bottleneck due to copying to cpu after every training iteration.

What exactly are you trying to do in your custom `ProgressBar`? Accumulating and averaging is a responsibility of our internal logging components, so I would suggest not trying to do this at the progress bar callback level.

**Comment by tobiasfshr** [comment]
I have a dictionary of scalar loss tensors that should be accumulated until the next printout happens in the progress bar, and then these should be averaged. If there is already accumulating and averaging happening in the internal components for all of progress_bar_metrics, could you point me to the piece of code? So far if i do not accumulate the values myself, the values seem to be way more unstable, so i assumed the values in trainer.progress_bar_metrics will be the latest ones

**Comment by carmocca** [comment]
Thanks, I understand what you are trying to do now.

The recommended approach, in that case, is to use a `torchmetrics.Metric` instance that you would update, `self.log("foo", avg_metric)` and you could access through your `LightningModule` to extract the `.compute()` value on-demand.

https://torchmetrics.readthedocs.io/en/stable/aggregation/mean.html

Here's a simple example:

```python
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer, Callback
from torchmetrics import MeanMetric


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)
        self.my_metric = MeanMetric()

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        self.my_metric.update(batch_idx + 1)
        self.log("my_metric", self.my_metric, prog_bar=True)
        return self(batch).sum()

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


class MySimpleProgressBar(Callback):
    def on_train_batch_end(self, trainer, pl_module, *args, **kwargs):
        print(pl_module.my_metric.compute())


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=3,
        max_epochs=1,
        enable_model_summary=False,
        enable_checkpointing=False,
        enable_progress_bar=False,
        logger=False,
        callbacks=MySimpleProgressBar()
    )
    trainer.fit(model, train_data)


if __name__ == "__main__":
    run()
```

**Comment by tobiasfshr** [comment]
okay thanks, that seems to work. However, this feels like a workaround since we are mixing logging responsibilities into the model. I'm not sure what the design goal is, if it is desirable to keep logging and model separate, but to me it seems doing this either in the progress bar or the LoggerConnector would be more intuitive. 
Plus one could actually log things dynamically, rather than having to define a variable for each in the __init__. This makes things easier e.g. when there is a loss that is only turned on after say a certain number of epochs, or in specific hyperparameter settings. I solved it like this for now:
```
    def training_step(self, batch, *args, **kwargs):
        losses = self.forward_train(batch)
        losses["loss"] = sum(list(losses.values()))

        log_dict = {}
        metric_attributes = []
        for k, v in losses.items():
            if not hasattr(self, k):
                metric = MeanMetric()
                metric.to(self.device)
                setattr(self, k, metric)

            metric = getattr(self, k)
            metric(v.detach())
            log_dict["train/" + k] = metric
            metric_attributes += [k]

        for (k, v), k_name in zip(log_dict.items(), metric_attributes):
            self.log(k, v, logger=True, prog_bar=True, on_step=True, on_epoch=False, metric_attribute=k_name)
        return losses
```
